pac4j-saml-config-db/src/main/java/org/pac4j/saml/dbclient/DbLoadedSamlClients.java

```diff
@@ -0,0 +1,197 @@
+import org.pac4j.core.client.Clients;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.util.CommonHelper;
+import org.pac4j.saml.client.SAML2Client;
+import org.pac4j.saml.dbclient.dao.api.SamlClientDao;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An extension to base {@link Clients}, which works also with clients whose definition is stored in a database. 
+ *
+ * @author jkacer
+ * @since 1.9.0
+ */
+public class DbLoadedSamlClients extends Clients {
+
+    /** SLF4J logger. */
+    private final Logger logger = LoggerFactory.getLogger(DbLoadedSamlClients.class);
+    
+    /** DAO reading definitions of SAML clients from a database. */
+    private SamlClientDao samlClientDao;
+
+    /** Clients set explicitly using a setter. */
+    private List<Client> explicitClients;
+    
+    // ------------------------------------------------------------------------------------------------------------------------------------
+
+    
+	/**
+	 * Creates a new client group with no explicit clients.
+	 */
+	public DbLoadedSamlClients() {
+		super();
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients.
+	 * 
+	 * @param clients
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(Client... clients) {
+		super(clients);
+	}
+
+
+	/**
+	 * Creates a new client group with a single explicit client.
+	 * 
+	 * @param client
+	 *            The client.
+	 */
+	public DbLoadedSamlClients(Client client) {
+		super(client);
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients.
+	 * 
+	 * @param clients
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(List<Client> clients) {
+		super(clients);
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients and a callback URL.
+	 * 
+	 * @param callbackUrl
+	 *            The callback URL.
+	 * @param clients
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(String callbackUrl, Client... clients) {
+		super(callbackUrl, clients);
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients and a callback URL.
+	 * 
+	 * @param callbackUrl
+	 *            The callback URL.
+	 * @param client
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(String callbackUrl, Client client) {
+		super(callbackUrl, client);
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients and a callback URL.
+	 *
+	 * @param callbackUrl
+	 *            The callback URL.
+	 * @param clients
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(String callbackUrl, List<Client> clients) {
+		super(callbackUrl, clients);
+	}
+
+
+	/**
+	 * Initializes all clients by loading them from the database.
+	 * 
+	 * @see org.pac4j.core.util.InitializableWebObject#internalInit(org.pac4j.core.context.WebContext)
+	 */
+	@Override
+	protected void internalInit(final WebContext context) {
+        // Remove all clients first if re-initialized

```

Comment:
Hi Jerome, I have finally kicked out the DbLoadedSamlClients class in commit 9014980. There should be no duplicate code anymore. What exactly I did:
- Modified the configuration as you suggested: init(name, context), loading properties from DB here, no longer an InitializableWebObject.
- DbLoadedSamlClients is gone --> no duplicate code anymore.
- In order to let Clients know about some clients configured in the database, I introduced interface AdditionalClientContributor and an implementation that scans available SAML clients defined in the database (just names, no details) and contributes them to Clients.
- The rest then works as you described: Clients finds a client (either static or dynamic) and the chosen client then initializes its configuration inside internalInit().

I could not find an easier way how to accomplish what I want to do.
Please review and share your comments. Thank you!